controller/controller_server/sync_lock.py

```diff
@@ -0,0 +1,54 @@
+import threading
+from collections import defaultdict
+
+from controller.common.csi_logger import get_stdout_logger
+from controller.controller_server.errors import ObjectAlreadyProcessingError
+
+logger = get_stdout_logger()
+
+
+class SyncLock:
+    def __init__(self):
+        self._lock = threading.Lock()
+        self._ids_in_use = defaultdict(set)
+        self.lock_key = ''
+        self.object_id = ''
+        self.action_name = ''
+
+    def add_to_ids_in_use(self, lock_key, object_id):
+        self._ids_in_use[lock_key].add(object_id)
+
+    def __call__(self, lock_key, object_id, action_name):
+        self.lock_key = lock_key
+        self.object_id = object_id
+        self.action_name = action_name
+        return self

```

Comment:
I don't think this is thread safe.
I think we need a `SyncLock` instance per `with` call, while sharing a "static" `_ids_in_use` between the instances

Reply:
done